python/pyspark/serializers.py

```diff
@@ -224,7 +225,13 @@ def _create_batch(series):
     # If a nullable integer series has been promoted to floating point with NaNs, need to cast
     # NOTE: this is not necessary with Arrow >= 0.7
     def cast_series(s, t):
-        if t is None or s.dtype == t.to_pandas_dtype():
+        if type(t) == pa.TimestampType:
+            # NOTE: convert to 'us' with astype here, unit ignored in `from_pandas` see ARROW-1680
+            return _series_convert_timestamps_internal(s).values.astype('datetime64[us]')
```

Comment:
It seems we need `s.fillna(0)` for null values.

Reply:
Why is that? We did that for integers that were promoted to floats to get rid of NaN, but here we are converting datetime64[ns] to datetime64[us] and both support missing values
```
In [28]: s = pd.Series([pd.datetime.now(), None])

In [29]: s
Out[29]: 
0   2017-10-24 10:44:51.483694
1                          NaT
dtype: datetime64[ns]

In [33]: s.values.astype('datetime64[us]')
Out[33]: array(['2017-10-24T10:44:51.483694', 'NaT'], dtype='datetime64[us]')
```

Reply:
I'm not exactly sure the reason but it seems `s.dt.tz_localize('tzlocal()')` in `_series_convert_timestamps_internal` doesn't work properly when including null values.

Reply:
hmmm, that's strange `s.dt.tz_localize('tzlocal()` gets an `OverflowError: Python int too large to convert to C long` error when printing but `s.dt.tz_localize('tzlocal()').dt.tz_convert('UTC')` works but comes up with a bogus time where the NaT was.  I agree that `fillna(0)` is safer to avoid overflow.

```
In [44]: s.dt.tz_localize('tzlocal()').dt.tz_convert('UTC')
Out[44]: 
0      2017-10-24 17:44:51.483694+00:00
1   1677-09-21 08:12:43.145224192+00:00
dtype: datetime64[ns, UTC]

Reply:
I fixed the date/time-related casting bugs in pyarrow and added new cast implementations -- conversions from one timestamp unit to another in Arrow-land fail silently right now, this will all be in the 0.8.0 release landing hopefully the week of 11/6 or thereabouts https://github.com/apache/arrow/pull/1245

Reply:
great, thanks @wesm!

Reply:
@BryanCutler It seems `s.apply(lambda ts: ts.tz_localize('tzlocal()'))` works without `s.fillna(0)`. Do you know the difference between this and `s.dt.tz_localize('tzlocal()')`?

Reply:
`apply()` will invoke the given function on each individual value of the series.  I think this iterates over the series, where `s.dt.tz_localize()` would do a vectorized operation and should be faster.